src/mapblock.cpp

```diff
@@ -24,6 +24,77 @@
 #include "util/serialize.h"
 #include "util/basic_macros.h"
 
+// Like a std::unordered_map<content_t, content_t>, but faster.
+//
+// Unassigned entries are marked with 0xFFFF.
+//
+// The static memory requires about 65535 * 2 bytes RAM in order to be
+// sure we can handle all content ids.
+class IdIdMapping
+{
+	static_assert(sizeof(content_t) == 2, "content_t must be 16-bit");
+
+private:
+	std::unique_ptr<content_t[]> m_mapping;
+	std::vector<content_t> m_dirty;
+
+	static thread_local std::unique_ptr<content_t[]> tl_mapping;
+
+public:
+	IdIdMapping()
+	{
+		m_mapping = std::move(tl_mapping);
+		if (!m_mapping) {
+			m_mapping = std::make_unique<content_t[]>(CONTENT_MAX + 1);
+			memset(m_mapping.get(), 0xFF, (CONTENT_MAX + 1) * sizeof(content_t));
+		}
+	}
+
+	DISABLE_CLASS_COPY(IdIdMapping)
+
+	IdIdMapping(IdIdMapping &&other) noexcept :
+		m_mapping(std::move(other.m_mapping)), m_dirty(std::move(other.m_dirty))
+	{
+	}
+
+	IdIdMapping &operator=(IdIdMapping &&other) noexcept
+	{
+		if (&other == this)
+			return *this;
+		m_mapping = std::move(other.m_mapping);
+		m_dirty = std::move(other.m_dirty);
+		return *this;
+	}
+
+	content_t get(content_t k)
+	{
+		return m_mapping[k];
+	}
+
+	void set(content_t k, content_t v)
+	{
+		m_mapping[k] = v;
+		m_dirty.push_back(k);
```

Comment:
keeping a single variable `m_dirty = std::max(m_dirty, v);` would probably be just as a fast in practice and avoids extra memory allocations

Reply:
For the local to global conversion, this should indeed be as good in practice, but for global to local, there can be huge gaps.

(If allocations are a concern, one could also make the dirty vector thread local. I haven't tried it.)

Reply:
Tried it. Made pretty much no difference (46 s with vector, 48 s with integer).

*Properly* testing would require me to make sure I include scenarios where there are many registered nodes and many mapblocks use content ids with high id value.

Reply:
Also tried to store dirty vector in a thread local variable. No gain.